Add external reference to Dossier and its index

tarnap opened the pull request:
Some institutions use an unambiguous reference field to find dossiers.
The composition of this field varies between the institutions using it
and therefore cannot be generated but is furnished by the user.

Overview of the made steps:
-  Zusätzliches TextLine Feld external_reference zum IDossier behavior hinzufügen.
-  Der Wert soll im SearchableText aufgenommen werden.
-  Darstelllung in der Byline falls Wert vorhanden
-  Update API Schema Docs using `bin/instange dump_schemas` and `bin/build-docs-public`
-  Update `README.rst`
-  Ein separater Index `external_reference` wurde hinzugefügt zum indexieren der `IDossier.external_reference` so wie `IDocumentMetadata.foreign_reference`. (Eine Abfrage ist vorerst nur via API oder GET-Parameter möglich. Siehe Screenshot in der Besprechung.)
- Tests für Volltextsuche so wie den `external_reference`-Index wurden hinzugefügt.

Resolves: #3080

phgross commented on opengever/dossier/viewlets/byline.py:
```diff
@@ -30,6 +30,10 @@ def mailto_link(self):
             return '<a href="mailto:%s">%s</a>' % (address, address)
 
+    def external_reference(self):
+        dossier = IDossier(self.context)
```
return IDossier(self.context). external_reference

phgross commented on opengever/dossier/tests/test_indexer.py:
```diff
@@ -125,6 +125,17 @@ def test_dossier_searchable_text_contains_keywords(self):
             index_data_for(dossier_with_keywords).get('SearchableText'))
 
+    def test_dossier_searchable_text_contains_external_reference(self):
+        dossier_with_external_reference = create(
```
dossier =

phgross commented on opengever/dossier/locales/de/LC_MESSAGES/opengever.dossier.po:
```diff
@@ -363,36 +363,40 @@ msgid "heading_dossier_note"
 
+#: ./opengever/dossier/behaviors/dossier.py:100
+msgid "help_external_reference"
+msgstr "Eindeutige Referenz zur Suche von Dossiers."
```
?? 🤔 

✂️

phgross commented on opengever/dossier/behaviors/dossier.py:
```diff
@@ -94,6 +95,12 @@ class IDossier(form.Schema):
 
+    external_reference = schema.TextLine(
+        title=_(u'label_external_reference', default=u'External Reference'),
+        description=_('help_external_reference', default=u''),
```
empty description --> ✂️

tarnap replied:
@phgross You want me to remove the help text completely?

phgross replied:
Yes, an empty field description has no value. As soon the customer wishes a description we can add one. 

Maybe you or @phabegger can ask via the FeedbackForum if a field description is necessary.

phgross commented on docs/public/dev-manual/api/schemas/opengever.dossier.businesscasedossier.inc:
```diff
@@ -31,7 +31,23 @@
+       :Datentyp: ``Text``
+
+
+       :Beschreibung: Indexiertes, eindeutiges Referenzzeichen zur Suche eines Dossiers. Dient als alternative Sicht auf das Dossier. Kann als Geschäftszeichen verwendet werden.
```
Where is that description from ?

tarnap replied:
I wrote it. I used the discussion on the feedback forum for this purpose.

phgross commented:
@tarnap please also add short overview of all the changes to the PR.

tarnap commented:
@phgross I made the changes, please re-review this pull request.

tarnap commented:
Upgrade step manually tested with @phgross 

To test the index a dossier was edited and an external reference was added:
![external_reference_add_dossier](https://user-images.githubusercontent.com/194114/29114162-ae8c7262-7cf3-11e7-94fa-72c70cf5160e.png)

This is the result for the lookup:
![external_reference_search_dossier](https://user-images.githubusercontent.com/194114/29114163-ae8e8534-7cf3-11e7-8ba1-c63efe85434d.png)

The same thing was done for a document and foreign reference:
![external_reference_add_document](https://user-images.githubusercontent.com/194114/29114414-86e95e54-7cf4-11e7-87d1-ff41f8394e64.png)
![external_reference_search_document](https://user-images.githubusercontent.com/194114/29114415-86ea6a60-7cf4-11e7-8c86-c2864def2909.png)

phgross commented:
@tarnap the description in the PR (`Overview of the made steps:`) is not up to date, please update

tarnap commented:
@tarnap Also the changelog entry is missing. Update that one as well

tarnap commented:
@phgross thanks for the reminder.
I added the entry to the changelog and added the steps to the overview in the pull request.
This pull request can be reviewed now.

phgross commented on opengever/core/upgrades/20170809094247_add_indexes_to_catalog/upgrade.py:
```diff
@@ -0,0 +1,9 @@
+from ftw.upgrade import UpgradeStep
+
+
+class AddIndexesToCatalog(UpgradeStep):
```
AddIndexesToCatalog is a bit to general
I would rename it to `AddExternalReferenceIndex`

phgross commented on opengever/document/tests/test_indexers.py:
```diff
@@ -94,6 +94,23 @@ def test_searchable_text(self):
             )
 
+    def test_external_reference(self):
+        doc1 = createContentInContainer(
```
Use `ftw.builder` to create test content in a FunctionalTest (or the fixture objects in IntegrationTest).

phgross commented on opengever/document/tests/test_indexers.py:
```diff
@@ -94,6 +94,23 @@ def test_searchable_text(self):
+            title=u'Doc Two',
+            foreign_reference=u'qpr-900-9002')
+
+        result = self.portal.portal_catalog(external_reference=u'qpr-900-9001')
```
I would not test searching the catalog. Just check if the right value is indexed.

```
from opengever.testing import index_data_for
self.assertEquals(
    u'Hugo Boss'
    index_data_for(doc1).get('external_reference'),
    )
```

phgross commented on opengever/dossier/tests/test_indexer.py:
```diff
@@ -148,6 +148,25 @@ def test_dossiertemplate_searchable_text_contains_keywords(self):
+            .having(external_reference=u'qpr-900-9002')
+            .within(self.repo_folder))
+
+        result = self.portal.portal_catalog(external_reference=u'qpr-900-9001')
```
Same here, assert on the indexed value

phgross commented on docs/HISTORY.txt:
```diff
@@ -18,6 +18,14 @@ Changelog
 - Improve warning when opening an old version of a document. [tarnap]
 - Highlight search terms after fetching new search results per ajax. [elioschmutz]
 - Fixed bug in search when sorting on relevance. [phgross]
+
```
I would say `- Add external_reference field and index [tarnap]` is enough.

The changelog can be short, all the details are visible in the PR/commits.

tarnap commented:
@phgross: I made the changes, please re-review
